File: src/cs/Ssh.Tcp/Messages/PortForwardSuccessMessage.cs
// Copyright (c) Microsoft. All rights reserved.

using Microsoft.DevTunnels.Ssh.IO;

namespace Microsoft.DevTunnels.Ssh.Messages;

public class PortForwardSuccessMessage : SessionRequestSuccessMessage
{
	public uint Port { get; set; }

	protected override void OnRead(ref SshDataReader reader)
	{
		base.OnRead(ref reader);

		// The port may be omitted from the response if it is the same as the requested port.
		if (reader.Available >= 4)
		{
			Port = reader.ReadUInt32();
		}
	}

	protected override void OnWrite(ref SshDataWriter writer)
	{
		base.OnWrite(ref writer);
		writer.Write(Port);
	}

	public override string ToString()
	{
		return base.ToString() + $"(Port: {Port})";
	}
}

File: src/ts/ssh-tcp/messages/portForwardSuccessMessage.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import {
	SessionRequestSuccessMessage,
	SshDataReader,
	SshDataWriter,
} from '@microsoft/dev-tunnels-ssh';

export class PortForwardSuccessMessage extends SessionRequestSuccessMessage {
	public port: number = 0;

	protected onRead(reader: SshDataReader): void {
		super.onRead(reader);

		if (reader.available >= 4) {
			this.port = reader.readUInt32();
		}
	}

	protected onWrite(writer: SshDataWriter): void {
		super.onWrite(writer);

		writer.writeUInt32(this.validateField(this.port, 'port'));
	}

	public toString() {
		return `${super.toString()} (port=${this.port})`;
	}
}
